### PR TITLE
Refactor HostManager and Provisioner to lessen locking in addNetwork

### LINKED ISSUE #2842: [runtime] Compiles are done under the HostManager networkLock
Currently the HostManager holds the networkLock_ while doing optimization, compile and provisioning of networks. This has two unfortunate consequences:
* Only one compile can be done at a time (across all devices).
* No executions can be started while a compile is occurring.

Fix this by introducing a lock per backend (or better yet, a queue) and making the provisioner more granularly thread safe.

### PR DESCRIPTION
**Pull request by gcatron**
Summary:
This PR refactors HostManager::addNetwork and the Provisioner. Compiling functions no longer happens under lock.
If a network is in the process of being added and a new add or remove call is made targeting the same network name an error is returned that the network specified is busy.

Documentation:

Fixes #2842 

Test Plan: Ninja check and verify tests pass.

### CODE REVIEW

File: include/glow/Runtime/HostManager/HostManager.h [hunk 1]
```diff
@@ -81,6 +81,9 @@ class HostManager final {
   /// onto the devices.
   std::unique_ptr<Provisioner> provisioner_;
 
+  /// Helper function to handle cleanup if an error occurs during addNetwork.
+  void cleanupAddNetwork(std::vector<std::string> names);
```
Comment (nadavrot) [inline include/glow/Runtime/HostManager/HostManager.h]: Please pass array_ref, or at least const-reference.

File: lib/Runtime/HostManager/HostManager.cpp [hunk 2]
```diff
@@ -155,21 +180,27 @@ llvm::Error HostManager::addNetwork(std::unique_ptr<Module> module,
     provisioner_.reset(new Provisioner(devices_));
     executor_.reset(new ThreadPoolExecutor(devices_, config_.executorThreads));
   }
-
-  RETURN_IF_ERR(provisioner_->provision(nodeList, *module, cctx));
-
+  {
+    // std::lock_guard<std::mutex> networkLock(networkLock_);
```
Comment (nadavrot) [inline lib/Runtime/HostManager/HostManager.cpp]: What is this?

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 3]
```diff
@@ -116,17 +143,27 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
           if (backends_[i]->getBackendName() == nodeBackendName) {
             auto compiledOrErr = backends_[i]->compile(function, options);
             if (!compiledOrErr) {
+              {
```
Comment (nadavrot) [inline lib/Runtime/Provisioner/Provisioner.cpp]: This code is not documented. Could you add some comments that explains the logic?

File: lib/Runtime/HostManager/HostManager.cpp [hunk 4]
```diff
@@ -92,21 +92,36 @@ HostManager::init(std::vector<std::unique_ptr<DeviceConfig>> configs) {
...
+      std::string name = F->getName();
+      auto it = networks_.find(name);
+      if (it != networks_.end()) {
+        cleanupAddNetwork(names);
+        return MAKE_ERR(
+            GlowErr::ErrorCode::RUNTIME_ERROR,
+            "Failed to add network: already have a function called " + name);
+      }
+      // Put an entry into networks to denote this network is being added.
+      networks_[name];
```
Comment (rdzhabarov) [inline lib/Runtime/HostManager/HostManager.cpp]: a bit weird, networks_.insert ?

File: tests/unittests/ProvisionerTest.cpp [hunk 5]
```diff
@@ -79,8 +79,8 @@ TEST_F(ProvisionerTest, provisionDag) {
   }
 
   CompilationContext cctx;
-  auto provisioner = Provisioner(devices);
```
Comment (rdzhabarov) [inline tests/unittests/ProvisionerTest.cpp]: curious why this change

Comment (gcatron) [reply tests/unittests/ProvisionerTest.cpp]: With the added mutex the default copy constructor is deleted, but a heap allocation isn't needed, changed back to stack allocation.

File: lib/Runtime/HostManager/HostManager.cpp [hunk 6]
```diff
@@ -92,21 +92,36 @@ HostManager::init(std::vector<std::unique_ptr<DeviceConfig>> configs) {
...
 HostManager::~HostManager() { llvm::toString(clearHost()); }
 
-llvm::Error HostManager::addNetwork(std::unique_ptr<Module> module,
-                                    CompilationContext &cctx,
-                                    bool saturateHost) {
-  std::lock_guard<std::mutex> networkLock(networkLock_);
-  auto functions = module->getFunctions();
-  for (auto &F : functions) {
-    std::string name = F->getName();
+void HostManager::cleanupAddNetwork(llvm::ArrayRef<std::string> names) {
```
Comment (rdzhabarov) [inline lib/Runtime/HostManager/HostManager.cpp]: I looked at the coverage for this PR: https://fb-glow-assets.s3.amazonaws.com/coverage/coverage-pull/3425/lib/Runtime/HostManager/HostManager.cpp.gcov.html

It looks like internals of the for loop below are not exercised. Can we make a test covering newly added code?

File: lib/Runtime/HostManager/HostManager.cpp (hunk 4 above)
Comment (nickgg) [reply lib/Runtime/HostManager/HostManager.cpp]: I think you actually want a different map here, `pendingNetworks_` or something - that way you can store the key without needing a dummy value that might accidentally get used.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 7]
```diff
@@ -61,16 +61,35 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
...
-  for (auto &network : networks) {
-    for (auto &node : network.nodes) {
-      for (auto logical : node->logicalDevices) {
-        auto it = logicalDevices.find(logical);
-        if (it != logicalDevices.end()) {
-          it->second.push_back(node.get());
-        } else {
-          logicalDevices.emplace(logical, std::vector<DAGNode *>{node.get()});
+  {
+    std::lock_guard<std::mutex> networkLock(functionsLock_);
```
Comment (nickgg) [inline lib/Runtime/Provisioner/Provisioner.cpp]: This makes me think the functionsLock_ protects `activeFunctions_` too since `functions_` is not touched in this block.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 8]
```diff
@@ -115,18 +142,31 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
         for (size_t i = 0, e = backends_.size(); i < e; i++) {
           if (backends_[i]->getBackendName() == nodeBackendName) {
             auto compiledOrErr = backends_[i]->compile(function, options);
+            // Check to see if an error was encountered while compiling.
             if (!compiledOrErr) {
+              // If and error occured, clean up provisioning state and return
+              // the error.
+              {
+                std::lock_guard<std::mutex> functionLock(functionsLock_);
+                cleanupProvision(localActiveNames);
```
Comment (nickgg) [inline lib/Runtime/Provisioner/Provisioner.cpp]: It looks to me as though you always grab the functionsLock_ before calling cleanupProvision, consider moving the lock into the method call to avoid missing it.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 9]
```diff
@@ -105,8 +124,16 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
     for (auto &node : device.second) {
       // Only compile if we haven't compiled before. If we have previously
       // compiled the function reuse it.
-      auto it = functions_.find(node->name);
-      if (it == functions_.end()) {
+      bool alreadyCompiled{false};
+      {
+        std::lock_guard<std::mutex> functionLock(functionsLock_);
```
Comment (nickgg) [inline lib/Runtime/Provisioner/Provisioner.cpp]: It's often more expensive to have a fine grained lock if you take and release it a lot, e.g. in this loop you take and release the lock 3 times per device?

I think you could refactor this to do the compile outside the lock then update all the state under the lock in one go.

File: lib/Runtime/Provisioner/Provisioner.cpp (hunk 7 above)
Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: It does, updated the comment to call this out.

File: lib/Runtime/HostManager/HostManager.cpp (hunk 6 above)
Comment (gcatron) [reply lib/Runtime/HostManager/HostManager.cpp]: @rdzhabarov  in that report blue means covered right? I think it's covered now.

File: lib/Runtime/Provisioner/Provisioner.cpp (hunk 9 above)
Comment (gcatron) [reply lib/Runtime/Provisioner/Provisioner.cpp]: Good call, I refactored this to pull the lock out of the loop entirely. Now all compiled functions are added to functions_ at the end all in one lock.

File: lib/Runtime/Provisioner/Provisioner.cpp [hunk 10]
```diff
@@ -61,23 +61,44 @@ llvm::Error Provisioner::provision(DAGListTy &networks, Module &module,
...
+          }
         }
       }
     }
   }
   if (cctx.backendOpts.collectConstants) {
     // @todo(nickg) must use VLOG_EVERY_N temporarily here due to old glog
     // version.
+    // Locking here because glog log every n is not thread safe.
+    std::lock_guard<std::mutex> functionsLock(functionsLock_);
```
Comment (nickgg) [inline lib/Runtime/Provisioner/Provisioner.cpp]: What about the VLOG_EVERY_N calls below?

File: lib/Runtime/HostManager/HostManager.cpp (hunk 6 above)
Comment (rdzhabarov) [reply lib/Runtime/HostManager/HostManager.cpp]: yup, blue is covered. red is not.
